### PR TITLE
Refactoring the JSONSchemaVisitor

### PR DESCRIPTION
**Pull request by w-vi**
- DRY primitive types
- Remove unnecessary heap allocations

### CODE REVIEW

File: src/refract/JSONSchemaVisitor.cc [hunk 1]
```diff
@@ -340,62 +336,22 @@ namespace refract
...
-            if (fixed && !e.empty()) {
-                ArrayElement *a = new ArrayElement;
-                StringElement *s = new StringElement;
-
-                s->set(*v);
-                a->push_back(s);
-                addMember("enum", a);
-            }
-        }
+        primitiveType(e, GetValue<StringElement>(e));
```
Comment (klokane) [inline src/refract/JSONSchemaVisitor.cc]: you know type sent into template, `GetValue<T>(e)` can be moved into `primitiveType<>()` body

File: src/refract/JSONSchemaVisitor.h [hunk 2]
```diff
@@ -36,8 +37,28 @@ namespace refract
...
         bool allItemsEmpty(const ArrayElement::ValueType* val);
 
+        template<typename T>
+        void setPrimitiveType(const T& e)
+        {
+            static_assert(sizeof(T) == 0, "Only String, Number, Boolean types allowed");
+        };
+
+        template<typename T>
+        void primitiveType(const T& e, const typename T::ValueType *v)
```
Comment (klokane) [inline src/refract/JSONSchemaVisitor.h]: Not sure, but I mean body of private template member function can be just "declared" in header and definition can be moved in to `cc`. we need no to expose it.

File: src/refract/JSONSchemaVisitor.h [hunk 3]
```diff
@@ -36,8 +37,28 @@ namespace refract
         void enumElement(const ArrayElement& e, const ArrayElement::ValueType *val);
         bool allItemsEmpty(const ArrayElement::ValueType* val);
 
+        template<typename T>
+        void setPrimitiveType(const T& e)
+        {
+            static_assert(sizeof(T) == 0, "Only String, Number, Boolean types allowed");
```
Comment (klokane) [inline src/refract/JSONSchemaVisitor.h]: :+1:

File: src/refract/JSONSchemaVisitor.cc [hunk 4]
```diff
@@ -340,62 +336,22 @@ namespace refract
 
     void JSONSchemaVisitor::visit(const NullElement& e)
     {
-        NullElement nil;
-        addMember("type", &nil);
+        addMember("type", new NullElement);
```
Comment (pksunkara) [inline src/refract/JSONSchemaVisitor.cc]: Is this heap allocation going to be cleaned up somewhere? I would prefer the original code.

File: src/refract/JSONSchemaVisitor.cc (hunk 1 above)
Comment (w-vi) [inline src/refract/JSONSchemaVisitor.cc]: I was getting an error something like `expected primary-expression before ‘>’ token` which I ahve to admit I don't understand .... need to dive deeper into templates :)

**Comment by w-vi** [comment]
@klokane  Are we doing C++03 or C++0x? As Circle CI is failing as there is nothing like `static_assert` which is part of of the C++0x standard.

**Comment by pksunkara** [comment]
We are C++03 since there is a limitation with nodejs in protagonist.

File: src/refract/JSONSchemaVisitor.cc (hunk 4 above)
Comment (w-vi) [inline src/refract/JSONSchemaVisitor.cc]: Original code was passing a pointer to a local variable allocated on stack ... and yes there is clean up, when the Visitor's desctructor is called.

Comment (klokane) [inline src/refract/JSONSchemaVisitor.cc]: original code is wrong, it return address of object allocated on stack.

**Comment by pksunkara** [comment]
Everything looks good. Please look into if you can improve the `GetValue` thing.

**Comment by klokane** [comment]
@w-vi 03, but I already thing about `static_assert<>` port into code base, there is a few point where it can be usefull instead of runtime check

**Comment by klokane** [comment]
@w-vi please look at https://github.com/apiaryio/drafter/blob/wvi/json-schema-cleanup/src/refract/JSONSchemaVisitor.cc#L95

there is
```
           if (...) {
                renderer.visit(*e.value.second);
            }
            else if (...) {
                renderer.setFixed(true);
                renderer.visit(*e.value.second);
            }
            else {
                renderer.visit(*e.value.second);
            }
```

It probably can be reduced into:
```
if (/*second case*/) {
  renderer.setFixed(true);
}
renderer.visit(*e.value.second);
```

**Comment by klokane** [comment]
https://github.com/apiaryio/drafter/blob/wvi/json-schema-cleanup/src/refract/JSONSchemaVisitor.cc#L156

please add explaining comment why you need no check availability of  `o1->value[0]`  it is not clear from where it come
``` 
                MemberElement *m1 = TypeQueryVisitor::as<MemberElement>(o1->value[0]->clone());
```

**Comment by w-vi** [comment]
@klokane So what to do about `static_assert`? Should I just put a FIXME there for now and put it back once we put `static_assert` in our codebase?

File: src/refract/JSONSchemaVisitor.cc (hunk 1 above)
Comment (klokane) [inline src/refract/JSONSchemaVisitor.cc]: `typename T::ValueType* v = GetValue<T>(e);`

**Comment by klokane** [comment]
@w-vi your decision. If you are willing to port `static_assert<>` go straight :)

**Comment by klokane** [comment]
But maybe leave it for now with FIXME. We need to discuss which one code base to integrate (drafter/snowcrash/refract).

**Comment by w-vi** [comment]
@klokane leaving the `static_assert` for later discussion and using the runtime approach

**Comment by klokane** [comment]
@w-vi this should be probably replaced by `TypeQueryVisitor::as<ObjectElement>` or is there reason to use check against string?
https://github.com/apiaryio/drafter/blob/bef8ac702e21337a97a0c2f762c6572e6d79f4a5/src/refract/JSONSchemaVisitor.cc#L208

**Comment by klokane** [comment]
All other seems ok

**Comment by w-vi** [comment]
@klokane no, `TypeQueryVisitor::as<ObjectElement>` is better, copy and paste error, as there is nothing like `EnumElement` and there I need to check the string `enum` .... let's do the `EnumElement` :)

**Comment by klokane** [comment]
Yeah `EnumElement` is already proposed. Hope for delivery as soon a possible :)

**Pull request merged**